src/bcf/split.rs

```diff
@@ -136,79 +74,33 @@ impl BreakendGroup {
     }
 }
 
-#[derive(Debug)]
-struct BcfInfo {
-    n_records: u64,
-    bnd_ends: HashMap<BreakendGroup, u64>,
-    supergroups: HashMap<BreakendGroup, BreakendGroup>,
-}
-
-impl BcfInfo {
-    fn new<P: AsRef<Path>>(input_bcf: P) -> Result<Self> {
-        let mut reader = bcf::Reader::from_path(input_bcf)?;
-        let mut bnd_ends = HashMap::new();
-        let mut supergroups: HashMap<BreakendGroup, BreakendGroup> = HashMap::new();
+type Id = Vec<u8>;
 
-        let mut i = 0;
-        let mut set_end = |group: BreakendGroup, i: u64| {
-            let end = bnd_ends.entry(group).or_insert(0);
-            *end = i;
-        };
-        for res in reader.records() {
-            let mut rec = res?;
-
-            if is_bnd(&mut rec) {
-                if let Some(group) = BreakendGroup::from(&mut rec) {
-                    if let Some(repr) = supergroups.get(&group) {
-                        // this group is part of a supergroup
-                        set_end(repr.clone(), i)
-                    } else {
-                        // register all subgroups
-                        for subgroup in group.subgroups() {
-                            if let Some(repr) = supergroups.get(&subgroup) {
-                                if repr.len() < group.len() {
-                                    supergroups.insert(subgroup, group.clone());
-                                }
-                            }
-                        }
-                        set_end(group, i)
-                    }
-                }
-            }
-            i += 1;
-        }
+trait BndRecord {
+    fn is_bnd(&self) -> bool;
+    fn event(&self) -> Option<Id>;
+    fn mateids(&self) -> Option<Vec<Id>>;
+}
 
-        Ok(BcfInfo {
-            n_records: i,
-            bnd_ends,
-            supergroups,
+impl BndRecord for bcf::Record {
+    fn is_bnd(&self) -> bool {
+        self.info(b"SVTYPE").string().map_or(false, |entries| {
+            entries.map_or(false, |entries| entries[0] == b"BND")
         })
     }
 
-    fn end(&self, group: &BreakendGroup) -> Option<u64> {
-        self.bnd_ends
-            .get(self.supergroups.get(group).unwrap_or(group))
-            .copied()
-    }
-}
-
-fn is_bnd(record: &mut bcf::Record) -> bool {
-    record.info(b"SVTYPE").string().map_or(false, |entries| {
-        entries.map_or(false, |entries| entries[0] == b"BND")
-    })
-}
-
-fn event(record: &mut bcf::Record) -> Option<Vec<u8>> {
-    if let Ok(Some(event)) = record.info(b"EVENT").string() {
-        Some(event[0].to_owned())
-    } else {
-        None
+    fn event(&self) -> Option<Id> {
+        if let Ok(Some(event)) = self.info(b"EVENT").string() {
+            Some(event[0].to_owned())
+        } else {
+            None
+        }
     }
-}
 
-fn mateids(record: &mut bcf::Record) -> Option<Vec<&[u8]>> {
-    match record.info(b"MATEID").string() {
-        Ok(Some(s)) => Some(s.clone().into_iter().collect_vec()),
-        _ => None,
```

Comment:
Note the difference to the new code which goes from `&[u8]` to `Vec<u8>`; the slice seems to have been "unstable" due to rust_htslib in the background so would sometimes be invalid (i.e. the memory it was pointing to changed!), which is why we allocate now.